fix: support varying batch sizes

stillmatic opened the pull request:
closes #653

I have signed the Microsoft CLA.

(apologies for the formatting changes, they were from running pre-commit black)

Commit:
feat: dynamic axes

Commit:
feat: add test for varying batch size

Commit:
chore: cleanup variable naming

Commit:
chore: lint

`pre-commit run --all-files` was quite aggressive

interesaaat commented on hummingbird/ml/_topology.py:
```diff
@@ -278,6 +283,7 @@ def convert(topology, backend, test_input, device, extra_config={}):
                 output_model_name,
                 input_names=topology.input_container.input_names,
                 output_names=topology.input_container.output_names,
+                dynamic_axes=dynamic_axes_cfg,
```
Should we remove the `fix_graph` code since this will make the trick?

Commit:
chore: remove `fix_graph`

interesaaat commented on tests/test_sklearn_gbdt_converter.py:
```diff
@@ -169,6 +169,28 @@ def test_float64_GB_trees_regressor_converter(self):
             self.assertIsNotNone(torch_model)
             np.testing.assert_allclose(model.predict(X), torch_model.predict(X), rtol=1e-06, atol=1e-06)
 
+    def test_varying_batch_sizes(self):
```
Should we get a similar test for `onnx`? This already worked before because torch uses dynamic dimensions by default

stillmatic replied:
nice catch, added logic checks in a few more places

Commit:
fix: check onnx conv, not torch

Commit:
chore: lint

Commit:
fix: skip onnx test if not installed

ksaur commented:
Thanks for working on this!! :)  I think the 30+ changed files are because our [line lengths](https://github.com/microsoft/hummingbird/blob/main/CONTRIBUTING.md#formatting) are usually 127 chars from the [flake8](https://github.com/microsoft/hummingbird/blob/main/.flake8).  (Although apparently it seems we have other various inconsistencies in our codebase heh.)  It might make it easier to troubleshoot, but no worries, not super strict!

Commit:
chore: add more tests

notably, the multi-classification models appear to still have the previous behavior.

Commit:
Revert "chore: lint"

This reverts commit 8b4ea4ed335a95b8a0fac7934b301d0b107173f4.

stillmatic commented:
thanks, I've reverted the formatting commit, I think my pre-commit isn't properly picking up the repo's config.

Commit:
fix: limit array properly

interesaaat commented on tests/test_onnxml_lightgbm_converter.py:
```diff
@@ -67,12 +67,16 @@ def _test_regressor(self, X, model, rtol=1e-06, atol=1e-06, extra_config={}):
-        np.testing.assert_allclose(
-            list(map(lambda x: list(x.values()), onnx_ml_pred[0])), onnx_pred[0], rtol=rtol, atol=atol
-        )  # probs
+            print("dfs", n, onnx_pred[0].shape, onnx_pred[0])
```
Can we remove this?

interesaaat commented on tests/test_onnxml_decision_tree_converter.py:
```diff
@@ -68,15 +68,18 @@ def _test_regressor(self, X, model, rtol=1e-06, atol=1e-06, extra_config={}):
+                rtol=rtol,
+                atol=atol,
+            )  # probs
+            print("predictions", onnx_pred)
```
Now that the tests pass can we remove the prints?

Commit:
chore: clean up remaining print statements

interesaaat commented on tests/test_onnxml_decision_tree_converter.py:
```diff
@@ -68,15 +68,17 @@ def _test_regressor(self, X, model, rtol=1e-06, atol=1e-06, extra_config={}):
-            rtol=rtol,
-            atol=atol,
-        )  # probs
+        for n in [1, len(X)]:
```
I don't think that this test is right. Should we convert the model once with X and then do predictions over the different batch sizes? As it is now this test just convert and run the model over the same batch size, it does not actual test that the sym dimension works.

stillmatic replied:
you're right, I've refactored the test harness to follow that logic.

Commit:
fix: set first dimension in tests to custom

Necessary to support arbitrary batch sizes
Also refactors tests to properly separate out model conversion and testing

Commit:
chore: update LGBM-ONNXML notebook

Must set first dimension to None in the ONNXMLTools conversion.

Commit:
fix: more bumpversion locations

stillmatic commented:
also preemptively pushing up a version bump - lmk if you'd rather make it a patch release (with a fix to the bumpversion config)

interesaaat commented:
> also preemptively pushing up a version bump - lmk if you'd rather make it a patch release (with a fix to the bumpversion config)

@ksaur WDYT?

ksaur commented:
I will take care of the patch release in a separate PR, thanks!

ksaur commented on hummingbird/ml/_topology.py:
```diff
@@ -289,43 +295,6 @@ def convert(topology, backend, test_input, device, extra_config={}):
-
-            return num_fixed
-
-        fix_graph(hb_model.graph)
```
This is exciting! 🎊

ksaur commented:
Thanks @stillmatic, revert the version bump and we'll merge!

stillmatic commented:
version bump is reverted!

interesaaat commented:
Thanks @stillmatic for fixing this!

ksaur commented:
Will push patch today by 5pmPT!
